app/controllers/concerns/course/assessment/submission/koditsu/submissions_concern.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+module Course::Assessment::Submission::Koditsu::SubmissionsConcern
+  extend ActiveSupport::Concern
+  include Course::Assessment::Question::KoditsuQuestionConcern
+  include Course::Assessment::Submission::Koditsu::UsersConcern
+  include Course::Assessment::Submission::Koditsu::AnswersConcern
+  include Course::Assessment::Submission::Koditsu::TestCasesConcern
+  include Course::Assessment::Submission::Koditsu::SubmissionTimesConcern
+
+  def fetch_all_submissions_from_koditsu(assessment, user)
+    @assessment = assessment
+    @user = user
+    submission_service = Course::Assessment::Submission::KoditsuSubmissionService.new(@assessment)
+    status, response = submission_service.run_fetch_all_submissions
+
+    return [status, nil] if status != 200 && status != 207
+
+    @all_submissions = response
+    @questions = @assessment.questions.includes({ actable: :test_cases })
+    @test_cases_order = test_cases_order_for(@questions)
+    @cu_submission_hash = course_user_submission_hash(@all_submissions)
+
+    process_all_submissions
+  end
+
+  private
+
+  def submission_status_hash
+    {
+      'inProgress' => 'attempting',
+      'submitted' => 'submitted'
+    }
+  end
+
+  def process_all_submissions
+    create_new_submissions_if_not_existing
+
+    @submission_hash = Course::Assessment::Submission.where(assessment: @assessment).to_h do |s|
+      [s.creator_id, s]
+    end
+
+    @cu_submission_hash.each do |creator, submission|
+      process_submission(submission, @submission_hash[creator.id])
+    end
+  end
+
+  def process_submission(submission, cm_submission)
+    state = submission_status_hash[submission['status']]
+    submitted_at = calculate_submission_time(state, submission['questions'])
+
+    cm_submission.class.transaction do
+      update_submission(cm_submission, state, submitted_at)
```

Comment:
Doesn't this still trigger N updates per submission?

Reply:
this is intentional, since we don't want the failure of fetching any submission to lead towards failure in fetching all previously fetched submissions (also it will be useful for debugging in the future, should the fetching fails in some point. Understand that ideally, fetching should be seamless and without fail, but since we're relying on third party API, this should be done).

Each submission's fetching should be handled separately

Reply:
In the first place I asked you to resolve N+1 queries, if you're not resolving N+1, you need to communicate that.  
Resolution for this is to put it into a job

Reply:
done